vpr/src/route/segment_stats.cpp

```diff
@@ -50,7 +50,7 @@ void get_segment_usage_stats(std::vector<t_segment_inf>& segment_inf) {
 
     for (size_t inode = 0; inode < device_ctx.rr_nodes.size(); inode++) {
         if (rr_graph.node_type(RRNodeId(inode)) == CHANX || rr_graph.node_type(RRNodeId(inode)) == CHANY) {
-            cost_index = device_ctx.rr_nodes[inode].cost_index();
+            cost_index = rr_graph.node_cost_index(RRNodeId(inode));
```

Comment:
Here's an example of where I think we should eventually hide the cost_index.
Instead of asking for a cost_index, and then looking up seg_type, we instead should have an api that takes an RRNodeId and returns the seg_type 
seg_type = rr_graph.seg_type(RRNodeId(inode));
Internally, the .seg_type() function would get the cost_index, and then look up the seg_type and return it.

It hides the fact that this seg_type lookup goes through a cost_index (flyweight) for storage efficiency; the caller doesn't care.

Reply:
It's possible there's some cpu time penalty to that in some code, if we ask for a bunch of cost_index data in rapid succession.  But in that case we could return a struct with the data that tends to be requested together packed into it.

Reply:
@vaughnbetz I think it is a good idea to streamline the API here. Actually, I am thinking about if we should include the ``rr_segments`` and ``rr_switch_inf`` as part of the ``RRGraphView`` and ``RRGraphBuilder``.

https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/77a3df829fdf509a76b9d7b941376f3835c2576a/vpr/src/base/vpr_context.h#L173-L177

I said so because these two data structures are modified copies of the original ones in architecture data. They have been updated to accurately model the switches and segments in the context of a specific routing resource graph.
Since they are adhoc to RRGraph, they can be part of the graph and we can provide shortcut APIs rather than force developers to visit several data structures.

However, it may be too complicated to this PR. It can be in a pull request near future (I can take the lead on it; since I have already tried in the [rr_graph_obj.h](https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/77a3df829fdf509a76b9d7b941376f3835c2576a/vpr/src/device/rr_graph_obj.h#L824))

Let me know what you think.

Reply:
If we decide to add API calls, the following are all a part of t_rr_indexed_data:     `base_cost`, `saved_base_cost`, `ortho_cost_index`, `seg_index`, `inv_length`, `T_linear`, `T_quadratic`, `C_load`.

The `ortho_cost_index` is used to access each of these data points for a different cost index, so we would have to consider how the API would handle that. We could have a second argument to the functions `rr_graph.seg_type(rr_node, is_ortho)` or potentially two separate functions `rr_graph.seg_type()`, `rr_graph.ortho_seg_type()`

Reply:
I agree, I think both those data structures should go in RRGraphView, and be built in RRGraphBuilder.
I also talked to @kmurray about this today in another meeting, and he also thinks hiding the cost_index and rr_indexed_data in RRGraphView api calls is the right way to go.

Reply:
I think the rr_graph.ortho_seg_type() style would be more clear.
For items that tend to be used together, I'd make the API return a tuple or struct with them packed in (e.g. inv_length, T_linear, T_quadratic and C_load are all probably used just for the classic lookahead and could be returned as a group only).